In [4]:
#Installing the MLxtend library in order to implement Apriori Algorithm
pip install mlxtend

Note: you may need to restart the kernel to use updated packages.


In [4]:
#Get our pandas and MLxtend code imported and read the data
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

df = pd.read_excel("C:/Users/Rodney/Downloads/Online_Retail.xlsx")
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

There is a little cleanup, we need to do. First, some of the descriptions have spaces that need to be removed. We’ll also drop the rows that don’t have invoice numbers and remove the credit transactions (those with invoice numbers containing C).

In [5]:
df['Description'] = df['Description'].str.strip()
df.dropna(axis = 0, subset = ['InvoiceNo'], inplace = True)
df['InvoiceNo'] = df['InvoiceNo'].astype('str')
df = df[~df['InvoiceNo'].str.contains('C')]

After the cleanup, we need to consolidate the items into 1 transaction per row with each product 1 hot encoded. For the sake of keeping the data set small, I’m only looking at sales for France. However, in additional code below, I will compare these results to sales from Germany. Further country comparisons would be interesting to investigate.

In [8]:
basket = (df[df['Country'] =="France"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

In [10]:
basket.head()

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536370                          0.0                         0.0   
536852                          0.0                         0.0   
536974                          0.0                         0.0   
537065                          0.0                         0.0   
537463                          0.0                         0.0   

Description  12 EGG HOUSE PAINTED WOOD  12 MESSAGE CARDS WITH ENVELOPES  \
InvoiceNo                                                                 
536370                             0.0                              0.0   
536852                             0.0                              0.0   
536974                             0.0                              0.0   
537065                             0.0                              0.0   
537463                             0.0                              0.0   

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                 0.0   
536852                                 0.0   
536974                                 0.0   
537065                                 0.0   
537463                                 0.0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536370                                       0.0                          0.0   
536852                                       0.0                          0.0   
536974                                       0.0                          0.0   
537065                                       0.0                          0.0   
537463                                       0.0                          0.0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536370                             0.0                                 0.0   
536852                             0.0                                 0.0   
536974                             0.0                                 0.0   
537065                             0.0                                 0.0   
537463                             0.0                                 0.0   

Description  12 PENCILS TALL TUBE WOODLAND  ...  WRAP VINTAGE PETALS  DESIGN  \
InvoiceNo                                   ...                                
536370                                 0.0  ...                          0.0   
536852                                 0.0  ...                          0.0   
536974                                 0.0  ...                          0.0   
537065                                 0.0  ...                          0.0   
537463                                 0.0  ...                          0.0   

Description  YELLOW COAT RACK PARIS FASHION  YELLOW GIANT GARDEN THERMOMETER  \
InvoiceNo                                                                      
536370                                  0.0                              0.0   
536852                                  0.0                              0.0   
536974                                  0.0                              0.0   
537065                                  0.0                              0.0   
537463                                  0.0                              0.0   

Description  YELLOW SHARK HELICOPTER  ZINC  STAR T-LIGHT HOLDER  \
InvoiceNo                                                         
536370                           0.0                        0.0   
536852                           0.0                        0.0   
536974                           0.0                        0.0   
537065                           0.0                        0.0   
537463                           0.0      

There are a lot of zeros in the data but we also need to make sure any positive values are converted to a 1 and anything less the 0 is set to 0. This step will complete the one hot encoding of the data and remove the postage column (since that charge is not one we wish to explore

In [11]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

basket_sets = basket.applymap(encode_units)
basket_sets.drop('POSTAGE', inplace = True, axis = 1)

Now that the data is structured properly, we can generate frequent item sets that have a support of at least 7% (this number was chosen so that I could get enough useful examples):

In [15]:
frequent_itemsets = apriori(basket_sets, min_support = 0.07, use_colnames = True)
frequent_itemsets.head()

support                        itemsets
0  0.071429   (4 TRADITIONAL SPINNING TOPS)
1  0.096939    (ALARM CLOCK BAKELIKE GREEN)
2  0.102041     (ALARM CLOCK BAKELIKE PINK)
3  0.094388      (ALARM CLOCK BAKELIKE RED)
4  0.081633  (BAKING SET 9 PIECE RETROSPOT)

The final step is to generate the rules with their corresponding support, confidence and lift:

In [16]:
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules

antecedents  \
0                        (ALARM CLOCK BAKELIKE GREEN)   
1                         (ALARM CLOCK BAKELIKE PINK)   
2                          (ALARM CLOCK BAKELIKE RED)   
3                        (ALARM CLOCK BAKELIKE GREEN)   
4                          (ALARM CLOCK BAKELIKE RED)   
5                         (ALARM CLOCK BAKELIKE PINK)   
6                                (SPACEBOY LUNCH BOX)   
7                              (DOLLY GIRL LUNCH BOX)   
8                          (PLASTERS IN TIN SPACEBOY)   
9                     (PLASTERS IN TIN CIRCUS PARADE)   
10                    (PLASTERS IN TIN CIRCUS PARADE)   
11                 (PLASTERS IN TIN WOODLAND ANIMALS)   
12                         (PLASTERS IN TIN SPACEBOY)   
13                 (PLASTERS IN TIN WOODLAND ANIMALS)   
14               (SET/20 RED RETROSPOT PAPER NAPKINS)   
15                      (SET/6 RED SPOTTY PAPER CUPS)   
16               (SET/20 RED RETROSPOT PAPER NAPKINS)   
17                    (SET/6 RED SPOTTY PAPER PLATES)   
18                    (SET/6 RED SPOTTY PAPER PLATES)   
19                      (SET/6 RED SPOTTY PAPER CUPS)   
20  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
21  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
22  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...   
23               (SET/20 RED RETROSPOT PAPER NAPKINS)   
24                    (SET/6 RED SPOTTY PAPER PLATES)   
25                      (SET/6 RED SPOTTY PAPER CUPS)   

                                          consequents  antecedent support  \
0                         (ALARM CLOCK BAKELIKE PINK)            0.096939   
1                        (ALARM CLOCK BAKELIKE GREEN)            0.102041   
2                        (ALARM CLOCK BAKELIKE GREEN)            0.094388   
3                          (ALARM CLOCK BAKELIKE RED)            0.096939   
4                         (ALARM CLOCK BAKELIKE PINK)            0.094388   
5                          (ALARM CLOCK BAKELIKE RED)            0.102041   
6                              (DOLLY GIRL LUNCH BOX)            0.125000   
7                                (SPACEBOY LUNCH BOX)            0.099490   
8                     (PLASTERS IN TIN CIRCUS PARADE)            0.137755   
9                          (PLASTERS IN TIN SPACEBOY)            0.168367   
10                 (PLASTERS IN TIN WOODLAND ANIMALS)            0.168367   
11                    (PLASTERS IN TIN CIRCUS PARADE)            0.170918   
12                 (PLASTERS IN TIN WOODLAND ANIMALS)            0.137755   
13                         (PLASTERS IN TIN SPACEBOY)            0.170918   
14                      (SET/6 RED SPOTTY PAPER CUPS)            0.132653   
15               (SET/20 RED RETROSPOT PAPER NAPKINS)            0.137755   
16                    (SET/6 RED SPOTTY PAPER PLATES)            0.132653   
17               (SET/20 RED RETROSPOT PAPER NAPKINS)            0.127551   
18                      (SET/6 RED SPOTTY PAPER CUPS)            0.127551   
19                    (SET/6 RED SPOTTY PAPER PLATES)            0.137755   
20                      (SET/6 RED SPOTTY PAPER CUPS)            0.102041   
21                    (SET/6 RED SPOTTY PAPER PLATES)            0.102041   
22               (SET/20 RED RETROSPOT PAPER NAPKINS)            0.122449   
23  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...            0.132653   
24  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...            0.127551   
25  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...            0.137755   

    consequent support   support  confidence      lift  leverage  conviction  
0             0.102041  0.073980    0.763158  7.478947  0.064088    3.791383  
1             0.096939  0.073980    0.725000  7.478947  0.064088    3.283859  
2             0.096939  0.079082    0.837838  8.642959  0.069932    5.568878  
3             0.094388  0.079082    0.815789  8.642959  0.069932    4.916181  
4             0.102041  0.0

Now, the tricky part is figuring out what this tells us. For instance, we can see that there are quite a few rules with a high lift value which means that it occurs more frequently than would be expected given the number of transaction and product combinations. We can also see several where the confidence is high as well. This part of the analysis is where the domain knowledge will come in handy. Since I do not have that, I’ll just look for a couple of illustrative examples.

We can filter the dataframe using standard pandas code. In this case, look for a large lift (6) and high confidence (.8):

In [17]:
rules[ (rules ['lift'] >= 6) &
      (rules ['confidence'] >= 0.8) ]      

antecedents  \
2                          (ALARM CLOCK BAKELIKE RED)   
3                        (ALARM CLOCK BAKELIKE GREEN)   
17                    (SET/6 RED SPOTTY PAPER PLATES)   
18                    (SET/6 RED SPOTTY PAPER PLATES)   
19                      (SET/6 RED SPOTTY PAPER CUPS)   
20  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
21  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
22  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...   

                             consequents  antecedent support  \
2           (ALARM CLOCK BAKELIKE GREEN)            0.094388   
3             (ALARM CLOCK BAKELIKE RED)            0.096939   
17  (SET/20 RED RETROSPOT PAPER NAPKINS)            0.127551   
18         (SET/6 RED SPOTTY PAPER CUPS)            0.127551   
19       (SET/6 RED SPOTTY PAPER PLATES)            0.137755   
20         (SET/6 RED SPOTTY PAPER CUPS)            0.102041   
21       (SET/6 RED SPOTTY PAPER PLATES)            0.102041   
22  (SET/20 RED RETROSPOT PAPER NAPKINS)            0.122449   

    consequent support   support  confidence      lift  leverage  conviction  
2             0.096939  0.079082    0.837838  8.642959  0.069932    5.568878  
3             0.094388  0.079082    0.815789  8.642959  0.069932    4.916181  
17            0.132653  0.102041    0.800000  6.030769  0.085121    4.336735  
18            0.137755  0.122449    0.960000  6.968889  0.104878   21.556122  
19            0.127551  0.122449    0.888889  6.968889  0.104878    7.852041  
20            0.137755  0.099490    0.975000  7.077778  0.085433   34.489796  
21            0.127551  0.099490    0.975000  7.644000  0.086474   34.897959  
22            0.132653  0.099490    0.812500  6.125000  0.083247    4.625850

In [18]:
basket['ALARM CLOCK BAKELIKE GREEN'].sum()

340.0

In [19]:
basket['ALARM CLOCK BAKELIKE RED'].sum()

316.0

In looking at the rules, it seems that the green and red alarm clocks are purchased together and the red paper cups, napkins and plates are purchased together in a manner that is higher than the overall probability would suggest.

At this point, you may want to look at how much opportunity there is to use the popularity of one product to drive sales of another. For instance, we can see that we sell 340 Green Alarm clocks but only 316 Red Alarm Clocks so maybe we can drive more Red Alarm Clock sales through recommendations?

What is also interesting is to see how the combinations vary by country of purchase. Let’s check out what some popular combinations might be in Germany:

In [20]:
basket2 = (df[df['Country'] =="Germany"]
          .groupby(['InvoiceNo', 'Description'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

basket_sets2 = basket2.applymap(encode_units)
basket_sets2.drop('POSTAGE', inplace=True, axis=1)
frequent_itemsets2 = apriori(basket_sets2, min_support=0.05, use_colnames=True)
rules2 = association_rules(frequent_itemsets2, metric="lift", min_threshold=1)

rules2[ (rules2['lift'] >= 4) &
        (rules2['confidence'] >= 0.5)]

antecedents                         consequents  \
0   (PLASTERS IN TIN CIRCUS PARADE)  (PLASTERS IN TIN WOODLAND ANIMALS)   
6        (PLASTERS IN TIN SPACEBOY)  (PLASTERS IN TIN WOODLAND ANIMALS)   
10    (RED RETROSPOT CHARLOTTE BAG)            (WOODLAND CHARLOTTE BAG)   

    antecedent support  consequent support   support  confidence      lift  \
0             0.115974            0.137856  0.067834    0.584906  4.242887   
6             0.107221            0.137856  0.061269    0.571429  4.145125   
10            0.070022            0.126915  0.059081    0.843750  6.648168   

    leverage  conviction  
0   0.051846    2.076984  
6   0.046488    2.011670  
10  0.050194    5.587746

It seems that in addition to David Hasselhoff, Germans love Plasters in Tin Spaceboy and Woodland Animals.